src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt

```diff
@@ -43,15 +41,38 @@ private constructor(
   private val chatPanel: ChatPanel = ChatPanel(project, this)
   private val cancellationToken = AtomicReference(CancellationToken())
   private val messages = mutableListOf<ChatMessage>()
+  private val newAgentSessionCallbacks =
+      Collections.synchronizedList(mutableListOf<Consumer<CodyAgent>>())
 
   init {
     cancellationToken.get().dispose()
-  }
+    doOnceAndOnNewAgentSession { agent ->
 
-  fun getPanel(): ChatPanel = chatPanel
+      // Register webview message listener on the agent.
```

Comment:
Agent connection is fully managed in the CodyAgentService already.
But `restoreAgentSession` is necessary to reload history of the every chat inside the agent (as agent is stateless, and if restarted looses any context).

The callback you added is different thing, and indded it would couple those object lifecycles a bit.
On the bright side I already resolved that issue by the way of the rewrite I did for multi-chat. 
Specifically, look at this part:
https://github.com/sourcegraph/jetbrains/blob/main/src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt#L24
There is new service called `AgentChatSessionService` which allow you to get chat session by sessionId/panelId. It's used to dispatch messages based on id to the correct `AgentChatSession` instance. Currently it calls `receiveMessage` on it, but you can add second callback (similar to `agent.client.onNewMessage`) which will in the same way dispatch to `receiveWebviewExtensionMessage`.